### PR TITLE
Special case of Ref and 0-dimensional AbstractArray in DataFrame and insertcols!

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2086

### CODE REVIEW

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -51,7 +51,9 @@ All columns in `columns` must be `AbstractVector`s and have the same length.
 An exception are `DataFrame(kwargs...)` and `DataFrame(pairs::Pair{Symbol,<:Any}...)`
 form constructors which additionally allow a column to be of any other type that is not
 an `AbstractArray`, in which case the passed value is automatically repeated to fill
-a new vector of the appropriate length.
+a new vector of the appropriate length. As a particular rule if a column is
+`Ref` or `0`-dimentional `AbstractArray` then it is allowed and in both cases
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: I'd just say "`Ref` or `0`-dimensional `AbstractArray` values are treated in the same way."

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -107,11 +109,21 @@ struct DataFrame <: AbstractDataFrame
...
             # recycle scalars
-            for i in 1:length(columns)
-                isa(columns[i], AbstractArray) && continue
-                columns[i] = fill!(Tables.allocatecolumn(typeof(columns[i]), maxlen),
-                                   columns[i])
-                lengths[i] = maxlen
+            for (i, col) in enumerate(columns)
+                if col isa AbstractArray
+                    if ndims(col) == 0
+                        x = col[]
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: You could probably avoid repeating the code by doing `if col isa Ref || (col isa AbstractArray and ndims(col) == 0)` and then `elseif !(col isa AbstractArray)`.

File: src/dataframe/dataframe.jl (hunk 1 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: fixed in both places

File: src/dataframe/dataframe.jl (hunk 2 above)
Comment (bkamins) [reply src/dataframe/dataframe.jl]: also fixed

**Comment by bkamins** [comment]
Thank you for the review.

File: src/dataframe/dataframe.jl [hunk 3]
```diff
@@ -51,7 +51,8 @@ All columns in `columns` must be `AbstractVector`s and have the same length.
 An exception are `DataFrame(kwargs...)` and `DataFrame(pairs::Pair{Symbol,<:Any}...)`
 form constructors which additionally allow a column to be of any other type that is not
 an `AbstractArray`, in which case the passed value is automatically repeated to fill
-a new vector of the appropriate length.
+a new vector of the appropriate length. As a particular rule values stored in
+`Ref` or `0`-dimentional `AbstractArray` are unwrapped and treated in the same way.
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]:
```suggestion
a `Ref` or a `0`-dimensional `AbstractArray` are unwrapped and treated in the same way.
```

File: src/dataframe/dataframe.jl [hunk 4]
```diff
@@ -596,7 +601,9 @@ If `ind` is omitted it is set to `ncol(df)+1`
 - `ind` : a position at which we want to insert a column
 - `name` : the name of the new column
 - `col` : an `AbstractVector` giving the contents of the new column or a value of any
-  type other than `AbstractArray` which will be repeated to fill a new vector
+  type other than `AbstractArray` which will be repeated to fill a new vector;
+  As a particular rule a values stored in `Ref` or `0`-dimentional `AbstractArray`
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]:
```suggestion
  As a particular rule a values stored in a `Ref` or a `0`-dimensional `AbstractArray`
```

File: src/dataframe/dataframe.jl [hunk 5]
```diff
@@ -685,8 +692,16 @@ function insertcols!(df::DataFrame, col_ind::Int, name_cols::Pair{Symbol,<:Any}.
 
     for (name, item) in name_cols
         if !(item isa AbstractVector)
-            @assert !(item isa AbstractArray)
-            item_new = fill!(Tables.allocatecolumn(typeof(item), target_row_count), item)
+            if item isa AbstractArray
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: How about merging the first two branches and doing `@assert item isa AbstractArray{<:Any, 0}`?

Comment (bkamins) [reply src/dataframe/dataframe.jl]: Good point. I merged them and made an assert in the following branch. Essentially we want to disallow `AbstractArray` of dimension larger than 1.

File: src/dataframe/dataframe.jl [hunk 6]
```diff
@@ -107,11 +108,15 @@ struct DataFrame <: AbstractDataFrame
         minlen, maxlen = extrema(lengths)
         if minlen != maxlen || minlen == maxlen == 1
             # recycle scalars
-            for i in 1:length(columns)
-                isa(columns[i], AbstractArray) && continue
-                columns[i] = fill!(Tables.allocatecolumn(typeof(columns[i]), maxlen),
-                                   columns[i])
-                lengths[i] = maxlen
+            for (i, col) in enumerate(columns)
+                if (col isa AbstractArray && ndims(col) == 0) || col isa Ref
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl]: Maybe use the same construct here?
```suggestion
                if col isa Union{AbstractArray{<:Any, 0}, Ref}
```

**Pull request merged**